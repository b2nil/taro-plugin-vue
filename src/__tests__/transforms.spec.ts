import {
  isMiniappNativeTag,
  transformEnv,
  transformH5Tags,
  transformH5AssetUrls,
  transformMiniappAssetUrls
} from '../transforms'

import {
  parse,
  // SFCTemplateBlock,
  compileTemplate,
  SFCTemplateCompileOptions
} from '@vue/compiler-sfc'
import { trackSlotScopes } from '@vue/compiler-core'

function compile (opts: Omit<SFCTemplateCompileOptions, 'id' | 'filename'>) {
  return compileTemplate({
    ...opts,
    filename: 'example.vue',
    id: 'whatever'
  })
}

function parseTemplate (source: string) {
  return parse(source, {
    filename: 'example.vue',
    sourceMap: true
  }).descriptor.template
}

const nativeTags = [
  "view", "icon", "progress", "rich-text", "text", "button", "checkbox", "checkbox-group",
  "form", "input", "label", "picker", "picker-view", "picker-view-column", "radio", "radio-group",
  "slider", "switch", "cover-image", "textarea", "cover-view", "movable-area", "movable-view",
  "scroll-view", "swiper", "swiper-item", "navigator", "audio", "camera", "image", "live-player",
  "video", "canvas", "ad", "web-view", "block", "map", "open-data", "custom-wrapper", "canvas",
  "editor", 'navigation-bar', 'official-account', 'functional-page-navigator'
]

describe('transformH5Tags', () => {
  test.concurrent.each(nativeTags)(`should resolve component 'taro-%s' `, async (tag: string) => {
    const result = compile({
      source: `<${tag} />`,
      compilerOptions: {
        nodeTransforms: [transformH5Tags()]
      }
    })
    expect(result.code).not.toMatch(`createElementBlock("${tag}")`)
    expect(result.code).toMatch(`createBlock(_component_taro_${tag.replace(/-/g, '_',)})`)
    expect(result.code).toMatch(`resolveComponent("taro-${tag}")`)
  })
})

describe('isMiniappNativeTag', () => {
  test.concurrent.each(nativeTags)('should treat mini-app tag `%s` as native tag', async (tag: string) => {
    const result = compile({
      source: `<${tag} />`,
      compilerOptions: {
        isNativeTag: isMiniappNativeTag
      }
    })

    expect(result.code).toMatch(`createElementBlock("${tag}")`)
    expect(result.code).not.toMatch(`resolveComponent("${tag}")`)
  })
})

describe('transformEnv', () => {
  const template = parseTemplate(`
<template>
<view taro-env="h5">taro-env=h5</view>
<view taro-env="h5" v-if="1!==2">taro-env=h5 v-if</view>
<view taro-env="h5" v-else>taro-env=h5 v-else</view>
<view taroEnv="h5">taroEnv=h5</view>
<view taroEnv="h5" v-for="i in [0, 1]" :key="i">taroEnv=h5-v-for-{{i}}</view>
<view taro-env="mini">taro-env=mini</view>
<view taro-env="mini" v-if="1!==2">taro-env=mini v-if</view>
<view taro-env="mini" v-else>taro-env=mini  v-else</view>
<view taroEnv="mini">taroEnv=mini</view>
<view taroEnv="mini" v-for="i in [0, 1]" :key="i">taroEnv=mini-v-for-{{i}}</view>
</template>
    `)

  test('should trasnform `taro-env="h5"`', () => {
    const resultH5 = compile({
      source: template!.content,
      compilerOptions: {
        nodeTransforms: [
          transformH5Tags(),
          transformEnv('h5')
        ]
      }
    })
    expect(resultH5.code).not.toMatch('taro-env=mini')
    expect(resultH5.code).not.toMatch('taroEnv=mini')
    expect(resultH5.code).toMatchSnapshot()
  })

  test('should trasnform `taro-env="mini"`', () => {
    const resultMini = compile({
      source: template!.content,
      compilerOptions: {
        isNativeTag: isMiniappNativeTag,
        nodeTransforms: [transformEnv()]
      }
    })

    expect(resultMini.code).not.toMatch('taro-env=h5')
    expect(resultMini.code).not.toMatch('taroEnv=h5')
    expect(resultMini.code).toMatchSnapshot()
  })
})

describe('transformAssetUrls', () => {
  test('should transform mini-app asset urls', () => {
    for (const [tag, urls] of Object.entries(transformMiniappAssetUrls)) {
      for (const url of urls) {
        const source = `<${tag} ${url}="~baz"/>`
        // Object option
        const { code: code1 } = compile({
          source,
          transformAssetUrls: {
            [tag]: [url]
          }
        })
        expect(code1).toMatch(`import _imports_0 from 'baz'\n`)
      }
    }
  })

  test('should transform h5 asset urls', () => {
    for (const [tag, urls] of Object.entries(transformH5AssetUrls)) {
      for (const url of urls) {
        const source = `<${tag} ${url}="~baz"/>`
        // Object option
        const { code: code1 } = compile({
          source,
          transformAssetUrls: {
            [tag]: [url]
          }
        })
        expect(code1).toMatch(`import _imports_0 from 'baz'\n`)
      }
    }
  })
})


