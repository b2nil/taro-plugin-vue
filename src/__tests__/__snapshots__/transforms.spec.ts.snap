// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformEnv should trasnform \`taro-env="h5"\` 1`] = `
"import { createTextVNode as _createTextVNode, resolveComponent as _resolveComponent, withCtx as _withCtx, createVNode as _createVNode, openBlock as _openBlock, createBlock as _createBlock, createCommentVNode as _createCommentVNode, renderList as _renderList, Fragment as _Fragment, createElementBlock as _createElementBlock, toDisplayString as _toDisplayString } from \\"vue\\"

const _hoisted_1 = /*#__PURE__*/_createTextVNode(\\"taro-env=h5\\")
const _hoisted_2 = /*#__PURE__*/_createTextVNode(\\"taro-env=h5 v-if\\")
const _hoisted_3 = /*#__PURE__*/_createTextVNode(\\"taro-env=h5 v-else\\")
const _hoisted_4 = /*#__PURE__*/_createTextVNode(\\"taroEnv=h5\\")

export function render(_ctx, _cache) {
  const _component_taro_view = _resolveComponent(\\"taro-view\\")

  return (_openBlock(), _createElementBlock(_Fragment, null, [
    _createVNode(_component_taro_view, null, {
      default: _withCtx(() => [
        _hoisted_1
      ]),
      _: 1 /* STABLE */
    }),
    (1!==2)
      ? (_openBlock(), _createBlock(_component_taro_view, { key: 0 }, {
          default: _withCtx(() => [
            _hoisted_2
          ]),
          _: 1 /* STABLE */
        }))
      : (_openBlock(), _createBlock(_component_taro_view, { key: 1 }, {
          default: _withCtx(() => [
            _hoisted_3
          ]),
          _: 1 /* STABLE */
        })),
    _createVNode(_component_taro_view, null, {
      default: _withCtx(() => [
        _hoisted_4
      ]),
      _: 1 /* STABLE */
    }),
    (_openBlock(), _createElementBlock(_Fragment, null, _renderList([0, 1], (i) => {
      return _createVNode(_component_taro_view, { key: i }, {
        default: _withCtx(() => [
          _createTextVNode(\\"taroEnv=h5-v-for-\\" + _toDisplayString(i), 1 /* TEXT */)
        ]),
        _: 2 /* DYNAMIC */
      }, 1024 /* DYNAMIC_SLOTS */)
    }), 64 /* STABLE_FRAGMENT */))
  ], 64 /* STABLE_FRAGMENT */))
}"
`;

exports[`transformEnv should trasnform \`taro-env="mini"\` 1`] = `
"import { renderList as _renderList, Fragment as _Fragment, openBlock as _openBlock, createElementBlock as _createElementBlock, createElementVNode as _createElementVNode, createCommentVNode as _createCommentVNode, toDisplayString as _toDisplayString } from \\"vue\\"

const _hoisted_1 = /*#__PURE__*/_createElementVNode(\\"view\\", null, \\"taro-env=mini\\", -1 /* HOISTED */)
const _hoisted_2 = { key: 0 }
const _hoisted_3 = { key: 1 }
const _hoisted_4 = /*#__PURE__*/_createElementVNode(\\"view\\", null, \\"taroEnv=mini\\", -1 /* HOISTED */)

export function render(_ctx, _cache) {
  return (_openBlock(), _createElementBlock(_Fragment, null, [
    _hoisted_1,
    (1!==2)
      ? (_openBlock(), _createElementBlock(\\"view\\", _hoisted_2, \\"taro-env=mini v-if\\"))
      : (_openBlock(), _createElementBlock(\\"view\\", _hoisted_3, \\"taro-env=mini v-else\\")),
    _hoisted_4,
    (_openBlock(), _createElementBlock(_Fragment, null, _renderList([0, 1], (i) => {
      return _createElementVNode(\\"view\\", { key: i }, \\"taroEnv=mini-v-for-\\" + _toDisplayString(i), 1 /* TEXT */)
    }), 64 /* STABLE_FRAGMENT */))
  ], 64 /* STABLE_FRAGMENT */))
}"
`;
